user/tests/wiring/no_fixture_power_saving/power_saving_mode.cpp

```diff
@@ -0,0 +1,235 @@
+#include "application.h"
+#include "test.h"
+
+#if HAL_PLATFORM_CELLULAR_LOW_POWER
+
+// Serial1LogHandler logHandler(115200, LOG_LEVEL_ALL);
+
+namespace {
+
+constexpr system_tick_t CLOUD_CONNECT_TIMEOUT = 10 * 60 * 1000;
+constexpr system_tick_t CLOUD_DISCONNECT_TIMEOUT = 1 * 60 * 1000;
+constexpr char skip_test_msg[] = "skip_test";
+int modem_type = DEV_UNKNOWN;
+String fnLp1Arg;
+
+int fnLp1(const String& arg) {
+    fnLp1Arg = arg;
+    return 12345;
+}
+
+} // namespace
+
+/**
+ * Returns current modem type:
+ * DEV_UNKNOWN, DEV_SARA_G350, DEV_SARA_U260, DEV_SARA_U270, DEV_SARA_U201, DEV_SARA_R410, DEV_SARA_R510
+ */
+int cellular_modem_type() {
+#if HAL_PLATFORM_NCP
+    CellularDevice device = {};
+    device.size = sizeof(device);
+    cellular_device_info(&device, NULL);
+
+    return device.dev;
+#endif // HAL_PLATFORM_NCP
+    return DEV_UNKNOWN;
+}
+
+uint32_t measureAvgPeriodMs(pin_t pin, uint32_t wait_ms) {
+    int high_pulse = 0;
+    int low_pulse = 0;
+    int count = 0;
+    // Dummy read to sync pulse edge
+    pulseIn(pin, HIGH);
+    pulseIn(pin, LOW);
+    uint32_t s = millis();
+    while (millis() - s < wait_ms) {
+        // This works better without an atomic block
+        // ATOMIC_BLOCK() {
+            high_pulse += pulseIn(pin, HIGH);
+            low_pulse += pulseIn(pin, LOW);
+        // }
+        count++;
+    }
+    high_pulse = high_pulse/count;
+    low_pulse = low_pulse/count;
+
+    return (high_pulse + low_pulse)/1000;
+}
+
+test(POWER_SAVING_00_setup) {
+#if HAL_PLATFORM_NCP
+    modem_type = cellular_modem_type();
+    if (modem_type != DEV_SARA_R510) {
+        pushMailbox(MailboxEntry().type(MailboxEntry::Type::DATA).data(skip_test_msg, sizeof(skip_test_msg) - 1));
+        skip();
+        return;
+    }
+#endif // HAL_PLATFORM_NCP
+
+    Particle.disconnect();
+    assertTrue(waitFor(Particle.disconnected, CLOUD_DISCONNECT_TIMEOUT));
+    Particle.function("fnlp1", fnLp1);
+}
+
+test(POWER_SAVING_01_particle_publish_publishes_an_event_after_low_power_active) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    bool publishResult = false;
+    bool lowPowerResult = false;
+    int attempts = 0;
+    uint32_t period = 0;
+    const uint32_t WAIT_FOR_LOW_POWER_ACTIVE_MS = 20000;
+    const uint32_t WAIT_FOR_LOW_POWER_MEAS_MS = 10000;
+    const uint32_t LOW_POWER_ATTEMPTS_MAX = 10; // XXX: extreme cases take up to 2.5 minutes to drop into low power mode
+    do {
+        Particle.connect();
+        assertTrue(waitFor(Particle.connected, CLOUD_CONNECT_TIMEOUT));
+        delay(WAIT_FOR_LOW_POWER_ACTIVE_MS); // wait for UPSV=1 default delay of ~9.2s before modem drops into low power mode idle mode.
```

Comment:
Does the connect() and delay() need to be inside the do/while loop? 
Or can we just do it once and then only check the state of the CTS1 pin in the loop?

Reply:
The idea behind adding it to the loop, was if we become disconnected while waiting for the low power state, it's less deterministic.  If it stays connected the whole time, there won't be an issue calling these again.